qa/L0_python_api/test_kserve.py

```diff
@@ -0,0 +1,234 @@
+        service = TestingUtils.setup_service(server, frontend)
+
+        TestingUtils.teardown_server(server)
+        TestingUtils.teardown_service(service)
+
+    @pytest.mark.parametrize("frontend, client_type", [HTTP_ARGS[:2], GRPC_ARGS[:2]])
+    def test_service_custom_port(self, frontend, client_type):
+        server = TestingUtils.setup_server()
+        options = frontend.Options(port=8005)
+        service = TestingUtils.setup_service(server, frontend, options)
+        client = TestingUtils.setup_client(client_type, url="localhost:8005")
+
+        # Confirms that service starts at port 8005
+        client.is_server_ready()
+
+        TestingUtils.teardown_client(client)
+        TestingUtils.teardown_service(service)
+        TestingUtils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend, client_type, url", [HTTP_ARGS, GRPC_ARGS])
+    def test_inference(self, frontend, client_type, url):
+        server = TestingUtils.setup_server()
+        service = TestingUtils.setup_service(server, frontend)
+
+        # TODO: use common/test_infer
+        assert TestingUtils.send_and_test_inference_identity(client_type, url=url)
+
+        TestingUtils.teardown_service(service)
+        TestingUtils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend, client_type, url", [HTTP_ARGS])
+    def test_http_req_during_shutdown(self, frontend, client_type, url):
+        server = TestingUtils.setup_server()
+        http_service = TestingUtils.setup_service(server, frontend)
+        http_client = client_type.InferenceServerClient(url=url)
+        model_name = "delayed_identity"
+        delay = 2  # seconds
+        input_data0 = np.array([[delay]], dtype=np.float32)
+
+        input0 = client_type.InferInput("INPUT0", input_data0.shape, "FP32")
+        input0.set_data_from_numpy(input_data0)
+
+        inputs = [input0]
+        outputs = [client_type.InferRequestedOutput("OUTPUT0")]
+
+        async_request = http_client.async_infer(
+            model_name=model_name, inputs=inputs, outputs=outputs
+        )
+        # http_service.stop() does not use graceful shutdown
+        TestingUtils.teardown_service(http_service)
+
+        # So, inference request will fail as http endpoints have been stopped.
+        with pytest.raises(InferenceServerException):
+            async_request.get_result(block=True, timeout=2)
+
+        # http_client.close() calls join() to terminate pool of greenlets
+        # However, due to an unsuccessful get_result(), async_request is still
+        # an active thread. Hence, join stalls until greenlet timeouts.
+        # Does not throw an exception, but displays error in logs.
+        TestingUtils.teardown_client(http_client)
+
+        # delayed_identity will still be an active model
+        # Hence, server.stop() causes InternalError: Timeout.
+        with pytest.raises(tritonserver.InternalError):
+            TestingUtils.teardown_server(server)
```

Comment:
Is this true? Why shutting down frontend causes model still marked as "in used"? CC @Tabrizian @kthui

Reply:
Is the delay more than shutdown timeout seconds?

Reply:
Delay was set to 2 seconds. Shutdown timeout is set to 30 seconds by default.

Reply:
Is it reproducible outside In-Process Python API?

Reply:
> Is this true? Why shutting down frontend causes model still marked as "in used"? CC @Tabrizian @kthui

Maybe some handshake like calling `InferenceRequestDelete` or a callback (request release, response complete, etc) done in the frontend would be missed if frontend is shutdown before responses are complete, and cause core to think the request is still active (and model in use, held by request)?